pypeit/core/scattlight.py

```diff
@@ -0,0 +1,308 @@
+    spat, spec = np.meshgrid(spatvec/(spatvec.size-1), specvec/(specvec.size - 1))
+    # Generate the polynomial efficiency scaling in the spatial direction
+    polyscale = spat*(polyterms_spat[0] + polyterms_spat[1]*spec)  # linear term and a spectral cross-term
+    # Now include the spectral direction
+    for pp in range(polyterms_spec.size):
+        polyscale += polyterms_spec[pp] * spec**pp
+
+    # Generate a 2D smoothing kernel, composed of a 2D Gaussian and a 2D Lorentzian
+    sigmx, sigmy = max(sigmx_g, sigmx_l), max(sigmy_g, sigmy_l),
+    xkern, ykern = np.meshgrid(np.arange(int(10 * sigmx)) - 5 * sigmx,
+                               np.arange(int(10 * sigmy)) - 5 * sigmy)
+    # Rotate the kernel
+    xkernrot = (xkern * np.cos(kern_angle) - ykern * np.sin(kern_angle))
+    ykernrot = (xkern * np.sin(kern_angle) + ykern * np.cos(kern_angle))
+    # Create and normalise the Gaussian kernel
+    kernel_gaussian = np.exp(-((xkernrot/sigmx_g)**2 + (ykernrot/sigmy_g)**2))
+    kernel_gaussian /= np.sum(kernel_gaussian)
+    # Create and normalise the Lorenztian kernel
+    kernel_lorentzian = 1 / ((xkernrot/sigmx_l) ** 2 + (ykernrot/sigmy_l) ** 2 + 1)
+    kernel_lorentzian /= np.sum(kernel_lorentzian)
+    # Add the individual kernels into a single kernel. Arbitrarily scale the Gaussian kernel (either is fine).
+    # The point of this is to make it so either a lorentzian, gaussian, or something in-between can be
+    # used as the kernel, making it more flexible for different spectrographs.
+    kernel = kernel_lorentzian + kern_scale * kernel_gaussian
+    kernel /= np.sum(kernel)
+
+    scale_img = polyscale * signal.oaconvolve(img, kernel, mode='same')
+    spl = interpolate.RectBivariateSpline(specvec, spatvec, scale_img, kx=1, ky=1)
+    return constant + spl(zoom_spec * (specvec + shft_spec), zoom_spat * (spatvec + shft_spat))
+
+
+def scattlight_resid(param, wpix, img):
+    """ Residual function used to optimize the model parameters
+
+    Parameters
+    ----------
+    param : `numpy.ndarray`_
+        1D array of model parameters to use for the fitting function.
+    wpix : tuple
+        A tuple containing the x,y coordinates of the pixels in img
+        to be used for computing the residual.
+    img : `numpy.ndarray`_
+        Data image to be used to compute the residual. Shape is (nspec, nspat)
+
+    Returns
+    -------
+    resid : `numpy.ndarray`_
+        A 1D vector of the residuals
+    """
+    model = scattered_light_model(param, img)
+    return img[wpix] - model[wpix]
+
+
+def scattered_light(frame, bpm, offslitmask, x0, bounds, detpad=300, debug=False):
+    """ Calculate a model of the scattered light of the input frame.
+
+    Parameters
+    ----------
+    frame : `numpy.ndarray`_
+        Raw 2D data frame to be used to compute the scattered light.
+    bpm : `numpy.ndarray`_
+        2D boolean array indicating the bad pixels (True=bad)
+    offslitmask : `numpy.ndarray`_
```

Comment:
Can you specify image shapes for frame and then state that bpm and offslitmask are the same size.